File: web/msp.js
const MSP = {
  API_VERSION: 1,
  FC_VARIANT: 2,
  FC_VERSION: 3,
  BOARD_INFO: 4,
  NAME: 10,
  STATUS: 101,
  RAW_IMU: 102,
  MOTOR: 104,
  RC: 105,
  ATTITUDE: 108,
  ANALOG: 110,
  BATTERY_STATE: 130,
  MOTOR_TELEMETRY: 139,
  STATUS_EX: 150,
};

const SENSOR_BITS = { gyro: 5, acc: 0, baro: 1, mag: 2, gps: 3, range: 4 };

const ARMING_DISABLE_FLAGS = [
  "NO_GYRO", "FAILSAFE", "RX_FAILSAFE", "NOT_DISARMED", "BOXFAILSAFE",
  "RUNAWAY", "CRASH", "THROTTLE", "ANGLE", "BOOT_GRACE", "NOPREARM",
  "LOAD", "CALIBRATING", "CLI", "CMS_MENU", "BST", "MSP", "PARALYZE",
  "GPS", "RESC", "RPMFILTER", "REBOOT_REQD", "DSHOT_BITBANG",
  "ACC_CALIB", "MOTOR_PROTO", "ARM_SWITCH",
];

function mspBuild(cmd, payload = new Uint8Array(0)) {
  const frame = new Uint8Array(6 + payload.length);
  frame.set([0x24, 0x4d, 0x3c, payload.length, cmd]);
  frame.set(payload, 5);
  let crc = 0;
  for (let i = 3; i < frame.length - 1; i++) crc ^= frame[i];
  frame[frame.length - 1] = crc;
  return frame;
}

class MspParser {
  constructor() {
    this.buf = [];
  }

  feed(chunk) {
    for (const b of chunk) this.buf.push(b);
    const frames = [];
    for (;;) {
      let start = -1;
      for (let i = 0; i + 1 < this.buf.length; i++) {
        if (this.buf[i] === 0x24 && this.buf[i + 1] === 0x4d) { start = i; break; }
      }
      if (start < 0) { this.buf.length = Math.min(this.buf.length, 1); break; }
      if (start > 0) this.buf.splice(0, start);
      if (this.buf.length < 5) break;
      const dir = this.buf[2];
      if (dir !== 0x3e && dir !== 0x21) { this.buf.splice(0, 2); continue; }
      const size = this.buf[3];
      if (this.buf.length < 6 + size) break;
      const cmd = this.buf[4];
      const payload = new Uint8Array(this.buf.slice(5, 5 + size));
      let crc = 0;
      for (let i = 3; i < 5 + size; i++) crc ^= this.buf[i];
      const ok = crc === this.buf[5 + size] && dir === 0x3e;
      this.buf.splice(0, 6 + size);
      if (ok) frames.push({ cmd, payload });
    }
    return frames;
  }
}

function dv(payload) {
  return new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
}

function parseStatus(p) {
  const out = {};
  if (p.length < 11) return out;
  const d = dv(p);
  out.cycle = d.getUint16(0, true);
  out.i2c = d.getUint16(2, true);
  const sensors = d.getUint16(4, true);
  out.sensors = {};
  for (const name in SENSOR_BITS) out.sensors[name] = !!((sensors >> SENSOR_BITS[name]) & 1);
  out.armed = !!(d.getUint32(6, true) & 1);
  if (p.length >= 13) out.cpuload = d.getUint16(11, true);
  if (p.length >= 20) {
    const flags = d.getUint32(16, true);
    out.arming = ARMING_DISABLE_FLAGS.filter((_, bit) => (flags >> bit) & 1);
  }
  return out;
}

function parseRawImu(p) {
  if (p.length < 18) return {};
  const d = dv(p);
  const v = [];
  for (let i = 0; i < 9; i++) v.push(d.getInt16(i * 2, true));
  return { acc: v.slice(0, 3).map((x) => x / 512), gyro: v.slice(3, 6) };
}

function parseAttitude(p) {
  if (p.length < 6) return {};
  const d = dv(p);
  return { att: [d.getInt16(0, true) / 10, d.getInt16(2, true) / 10, d.getInt16(4, true)] };
}

function parseAnalog(p) {
  if (p.length < 7) return {};
  const d = dv(p);
  const out = {
    vbat: d.getUint8(0) / 10,
    mah: d.getUint16(1, true),
    rssi: Math.round((d.getUint16(3, true) * 100) / 1023),
    amps: d.getInt16(5, true) / 100,
  };
  if (p.length >= 9) out.vbat = d.getUint16(7, true) / 100;
  return out;
}

function parseBatteryState(p) {
  if (p.length < 8) return {};
  const d = dv(p);
  const out = {
    cells: d.getUint8(0),
    capacity: d.getUint16(1, true),
    vbat: d.getUint8(3) / 10,
    mah: d.getUint16(4, true),
    amps: d.getInt16(6, true) / 100,
  };
  if (p.length >= 11) out.vbat = d.getUint16(9, true) / 100;
  return out;
}

function parseMotor(p) {
  const d = dv(p);
  const motors = [];
  for (let i = 0; i + 1 < p.length; i += 2) {
    const v = d.getUint16(i, true);
    if (v > 0) motors.push(v);
  }
  return { motors };
}

function parseMotorTelemetry(p) {
  if (!p.length) return {};
  const d = dv(p);
  const count = d.getUint8(0);
  const telem = [];
  let off = 1;
  for (let i = 0; i < count && off + 13 <= p.length; i++, off += 13) {
    telem.push({
      rpm: d.getUint32(off, true),
      temp: d.getUint8(off + 6),
      amps: d.getUint16(off + 9, true) / 100,
    });
  }
  return { telem };
}

function parseRc(p) {
  const d = dv(p);
  const channels = [];
  for (let i = 0; i + 1 < p.length; i += 2) channels.push(d.getUint16(i, true));
  return { channels };
}

function cleanAscii(p) {
  let out = "";
  for (const b of p) if (b >= 32 && b < 127) out += String.fromCharCode(b);
  return out.trim();
}

const MSP_PARSERS = {
  [MSP.STATUS]: parseStatus,
  [MSP.STATUS_EX]: parseStatus,
  [MSP.RAW_IMU]: parseRawImu,
  [MSP.ATTITUDE]: parseAttitude,
  [MSP.ANALOG]: parseAnalog,
  [MSP.BATTERY_STATE]: parseBatteryState,
  [MSP.MOTOR]: parseMotor,
  [MSP.MOTOR_TELEMETRY]: parseMotorTelemetry,
  [MSP.RC]: parseRc,
};

File: packaging/aur/src/fpvtop-0.2.0/fpvtop/msp.py
import struct
import time

MSP_API_VERSION = 1
MSP_FC_VARIANT = 2
MSP_FC_VERSION = 3
MSP_BOARD_INFO = 4
MSP_NAME = 10
MSP_STATUS = 101
MSP_RAW_IMU = 102
MSP_MOTOR = 104
MSP_RC = 105
MSP_ATTITUDE = 108
MSP_ANALOG = 110
MSP_BATTERY_STATE = 130
MSP_MOTOR_TELEMETRY = 139
MSP_STATUS_EX = 150
MSP_UID = 160

BAUD_RATE = 115200

SENSOR_BITS = {"gyro": 5, "acc": 0, "baro": 1, "mag": 2, "gps": 3, "range": 4}

ARMING_DISABLE_FLAGS = [
    "NO_GYRO", "FAILSAFE", "RX_FAILSAFE", "NOT_DISARMED", "BOXFAILSAFE",
    "RUNAWAY", "CRASH", "THROTTLE", "ANGLE", "BOOT_GRACE", "NOPREARM",
    "LOAD", "CALIBRATING", "CLI", "CMS_MENU", "BST", "MSP", "PARALYZE",
    "GPS", "RESC", "RPMFILTER", "REBOOT_REQD", "DSHOT_BITBANG",
    "ACC_CALIB", "MOTOR_PROTO", "ARM_SWITCH",
]


def build(cmd, payload=b""):
    frame = bytes([len(payload), cmd]) + payload
    crc = 0
    for byte in frame:
        crc ^= byte
    return b"$M<" + frame + bytes([crc])


class Parser:
    def __init__(self):
        self.buf = bytearray()

    def feed(self, data):
        self.buf.extend(data)
        frames = []
        while True:
            start = self.buf.find(b"$M")
            if start < 0:
                self.buf.clear()
                break
            if start > 0:
                del self.buf[:start]
            if len(self.buf) < 5:
                break
            direction = self.buf[2]
            if direction not in (0x3E, 0x21):
                del self.buf[:2]
                continue
            size = self.buf[3]
            total = 6 + size
            if len(self.buf) < total:
                break
            cmd = self.buf[4]
            payload = bytes(self.buf[5:5 + size])
            crc = 0
            for byte in self.buf[3:5 + size]:
                crc ^= byte
            ok = crc == self.buf[5 + size] and direction == 0x3E
            del self.buf[:total]
            frames.append((ok, cmd, payload))
        return frames


class Link:
    def __init__(self, port):
        import serial
        self.ser = serial.Serial()
        self.ser.port = port
        self.ser.baudrate = BAUD_RATE
        self.ser.timeout = 0.05
        self.ser.write_timeout = 0.5
        try:
            self.ser.exclusive = True
        except (AttributeError, ValueError):
            pass
        self.ser.open()
        time.sleep(0.1)
        self.ser.reset_input_buffer()
        self.parser = Parser()

    def close(self):
        try:
            self.ser.close()
        except Exception:
            pass

    def request(self, cmd, payload=b"", timeout=0.35):
        self.ser.write(build(cmd, payload))
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            chunk = self.ser.read(256)
            if not chunk:
                continue
            for ok, rcmd, rpayload in self.parser.feed(chunk):
                if rcmd == cmd:
                    return rpayload if ok else None
        return None


def find_port():
    try:
        from serial.tools import list_ports
    except ImportError:
        return None
    candidates = []
    for info in list_ports.comports():
        if info.vid == 0x0483 and info.pid == 0x5740:
            return info.device
        if "ACM" in info.device or "usbmodem" in info.device:
            candidates.append(info.device)
    return candidates[0] if candidates else None


def parse_status(payload):
    out = {}
    if len(payload) >= 11:
        cycle, i2c, sensors, mode, profile = struct.unpack_from("<HHHIB", payload, 0)
        out.update(cycle=cycle, i2c=i2c, profile=profile,
                   sensors={name: bool(sensors >> bit & 1) for name, bit in SENSOR_BITS.items()},
                   armed=bool(mode & 1))
    if len(payload) >= 13:
        out["cpuload"] = struct.unpack_from("<H", payload, 11)[0]
    if len(payload) >= 20:
        flags = struct.unpack_from("<I", payload, 16)[0]
        out["arming"] = [name for bit, name in enumerate(ARMING_DISABLE_FLAGS) if flags >> bit & 1]
    return out


def parse_raw_imu(payload):
    if len(payload) < 18:
        return {}
    values = struct.unpack_from("<9h", payload, 0)
    return {
        "acc": tuple(v / 512.0 for v in values[0:3]),
        "gyro": tuple(float(v) for v in values[3:6]),
        "mag": values[6:9],
    }


def parse_attitude(payload):
    if len(payload) < 6:
        return {}
    roll, pitch, yaw = struct.unpack_from("<hhh", payload, 0)
    return {"att": (roll / 10.0, pitch / 10.0, float(yaw))}


def parse_analog(payload):
    if len(payload) < 7:
        return {}
    legacy_v, mah, rssi, amps = struct.unpack_from("<BHHh", payload, 0)
    vbat = legacy_v / 10.0
    if len(payload) >= 9:
        vbat = struct.unpack_from("<H", payload, 7)[0] / 100.0
    return {"vbat": vbat, "mah": mah, "rssi": round(rssi * 100 / 1023), "amps": amps / 100.0}


def parse_battery_state(payload):
    if len(payload) < 8:
        return {}
    cells, capacity, legacy_v, mah, amps = struct.unpack_from("<BHBHh", payload, 0)
    out = {"cells": cells, "capacity": capacity, "mah": mah, "amps": amps / 100.0,
           "vbat": legacy_v / 10.0}
    if len(payload) >= 9:
        out["batt_state"] = payload[8]
    if len(payload) >= 11:
        out["vbat"] = struct.unpack_from("<H", payload, 9)[0] / 100.0
    return out


def parse_motor(payload):
    count = len(payload) // 2
    values = struct.unpack_from("<%dH" % count, payload, 0)
    return {"motors": [v for v in values if v > 0]}


def parse_motor_telemetry(payload):
    if not payload:
        return {}
    count = payload[0]
    telem = []
    offset = 1
    for _ in range(count):
        if offset + 13 > len(payload):
            break
        rpm, invalid, temp, volts, amps, mah = struct.unpack_from("<IHBHHH", payload, offset)
        telem.append({"rpm": rpm, "invalid": invalid / 100.0, "temp": temp,
                      "volts": volts / 100.0, "amps": amps / 100.0, "mah": mah})
        offset += 13
    return {"telem": telem}


def parse_rc(payload):
    count = len(payload) // 2
    return {"channels": list(struct.unpack_from("<%dH" % count, payload, 0))}


def clean_ascii(raw):
    return "".join(chr(b) for b in raw if 32 <= b < 127).strip()


def read_ident(link):
    ident = {}
    payload = link.request(MSP_API_VERSION)
    if payload and len(payload) >= 3:
        ident["api"] = "%d.%d" % (payload[1], payload[2])
    payload = link.request(MSP_FC_VARIANT)
    if payload:
        ident["variant"] = clean_ascii(payload)
    payload = link.request(MSP_FC_VERSION)
    if payload and len(payload) >= 3:
        ident["version"] = "%d.%d.%d" % (payload[0], payload[1], payload[2])
    payload = link.request(MSP_BOARD_INFO)
    if payload and len(payload) >= 4:
        ident["board"] = clean_ascii(payload[:4])
    payload = link.request(MSP_NAME)
    if payload is not None:
        ident["name"] = clean_ascii(payload)
    return ident
